inference-engine/src/auto_plugin/auto_plugin.cpp

```diff
@@ -0,0 +1,206 @@
+        IE_THROW() << "Please, work with AUTO device via InferencEngine::Core object";
+    }
+
+    if (network.getFunction() == nullptr) {
+        IE_THROW() << "AUTO device supports just ngraph network representation";
+    }
+
+    auto fullConfig = mergeConfigs(_config, config);
+    auto metaDevices = GetDeviceChoice(fullConfig);
+
+    // FIXME: always select CPU device now
+    DeviceInformation selectedDevice = SelectDevice(metaDevices);
+    IE::ExecutableNetwork executableNetwork;
+    try {
+        executableNetwork = GetCore()->LoadNetwork(network, selectedDevice.deviceName, selectedDevice.config);
+    } catch(const IE::Exception &iie) {
+        IE_THROW() << "Failed to load network to device named " << selectedDevice.deviceName;
+    }
+
+    bool enablePerfCounters = false;
+    try {
+        enablePerfCounters =
+            executableNetwork.GetConfig(IE::PluginConfigParams::KEY_PERF_COUNT).as<std::string>() ==
+                IE::PluginConfigParams::YES;
+    } catch (...) {
+    }
+
+    return std::make_shared<AutoExecutableNetwork>(executableNetwork,
+                                                   selectedDevice,
+                                                   enablePerfCounters);
+}
+
+IE::QueryNetworkResult AutoInferencePlugin::QueryNetwork(const IE::CNNNetwork& network, const ConfigType& config) const {
+    IE::QueryNetworkResult queryResult;
+    if (GetCore() == nullptr) {
+        IE_THROW() << "Please, work with AUTO device via InferencEngine::Core object";
+    }
+
+    if (network.getFunction() == nullptr) {
+        IE_THROW() << "AUTO device supports just ngraph network representation";
+    }
+
+    queryResult.rc = IE::StatusCode::OK;
+    queryResult.supportedLayersMap.clear();
+
+    auto fullConfig = mergeConfigs(_config, config);
+    auto metaDevices = GetDeviceChoice(fullConfig);
+    std::unordered_set<std::string> supportedLayers;
+    std::unordered_set<std::string> supportedDevices;
+    for (auto&& value : metaDevices) {
+        try {
+            if (value.deviceName != _pluginName) {
+                auto deviceQr = GetCore()->QueryNetwork(network, value.deviceName, value.config);
+                std::unordered_set<std::string> deviceSupportedLayers;
+                for (auto &&layerQr : deviceQr.supportedLayersMap) {
+                    deviceSupportedLayers.emplace(layerQr.first);
+                }
+                supportedLayers = supportedLayers.empty()
+                                ? deviceSupportedLayers : (deviceSupportedLayers.empty()
+                                ? supportedLayers : IE::details::Intersection(
+                                     supportedLayers, deviceSupportedLayers));
+                supportedDevices.insert(value.deviceName);
+            }
+        } catch (...) {
+        }
+    }
+
+    if (supportedDevices.empty()) {
+        IE_THROW() << "Please, check environment due to no supported devices can be used";
+    }
```

Comment:
may be this exception should be a part of `GetDeviceChoice `?

Reply:
done